# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/BrowserHost.cpp
# Commit: 6709a3bc1a1c
# Full Hash: 6709a3bc1a1c4ac18dce0f6db78efe12b3e1f178
# Author: Kashav Madan <kmadan@mozilla.com>
# Date: 2019-07-15 21:39:46
# Regressor Bug: 1543251
# File Overlap Count: 1
# Description:
#   Bug 1543251 - Move hasBeforeUnload from PBrowser to PWindowGlobal, r=nika
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D37003
# ==============================================================================

diff -r 83881815f007 -r 6709a3bc1a1c dom/ipc/BrowserHost.cpp
--- a/dom/ipc/BrowserHost.cpp	Mon Jul 15 11:01:04 2019 +0000
+++ b/dom/ipc/BrowserHost.cpp	Mon Jul 15 17:30:26 2019 +0000
@@ -250,15 +250,22 @@
 /* readonly attribute boolean hasBeforeUnload; */
 NS_IMETHODIMP
 BrowserHost::GetHasBeforeUnload(bool* aHasBeforeUnload) {
-  if (!mRoot) {
+  if (!mRoot || !GetBrowsingContext()) {
     *aHasBeforeUnload = false;
     return NS_OK;
   }
+
   bool result = false;
 
-  VisitAll([&result](BrowserParent* aBrowserParent) {
-    result |= aBrowserParent->GetHasBeforeUnload();
-  });
+  GetBrowsingContext()->PreOrderWalk(
+      [&result](BrowsingContext* aBrowsingContext) {
+        WindowGlobalParent* windowGlobal =
+            aBrowsingContext->Canonical()->GetCurrentWindowGlobal();
+
+        if (windowGlobal) {
+          result |= windowGlobal->HasBeforeUnload();
+        }
+      });
 
   *aHasBeforeUnload = result;
   return NS_OK;